Fixes #25921 - Sync Smart Proxies before installing Errata during Incremental Update

jturel opened the pull request:
Builds on the changes in #7946 and addresses the additional feedback here: https://github.com/Katello/katello/pull/7946#issuecomment-506422122

The goal of this change is to sync affected smart proxies before installing applicable errata during an incremental update. It achieves this by providing a new execution path through the `PromoteToEnvironment` action which will plan the smart proxy sync in the plan phase rather than using .async_task in the run phase. The behavior of CV promotion is not changed outside of incremental update.

I had a very bad time trying to test this on master with pulp3. I tried switching back to pulp2 but experienced different issues but those may have been related for my environment. For those reasons I would recommend testing this against 3.15 where I found it to work flawlessly.

**To test**

- Create a new Lifecycle Environment, Library -> **Stage**
- register a proxy to Katello
- associate Library and Stage environments to the proxy
- register a client (katello-client) to the proxy and make sure katello-agent is running.
- create a product with a repo that provides some errata: https://jlsherrill.fedorapeople.org/fake-repos/needed-errata/
- create a content view with the repo you created
- Promote the CV to Stage
- Add the client to the Stage env and the CV you created
- Install walrus-0.71 on the client
- add an errata excludes filter to the CV to omit the 'sea' errata in the repo. publish and promote it to Stage
- walrus-5.2 should **not** be available to the client (verify)

**Finally:**

- initiate an incremental update of the CVV you published and promoted: `hammer content-view version incremental-update  --content-view-version-id=$LATEST_CVV_ID --errata-ids=RHEA-2012:0055 --update-all-hosts=1 --lifecycle-environment-ids=$STAGE_LCE_ID
`
- verify walrus-5.2 is now install on the client

jturel commented on test/actions/katello/content_view_test.rb:
```diff
@@ -416,5 +417,37 @@ class IncrementalUpdatesTest < TestBase
+      hosts = ::Host.all
+
+      planned = plan_action(action, cvv_envs, [], content, false, hosts, "", true)
+      run_action(planned)
```
This test isn't fully built out yet but it's currently failing with this, and not sure what to make of it:

```
  1) Error:
Actions::Katello::ContentView::IncrementalUpdatesTest#test_0005_runs:
NoMethodError: undefined method `adapter' for nil:NilClass
    /home/vagrant/katello/test/actions/katello/content_view_test.rb:450:in `block in <class:IncrementalUpdatesTest>'
```

No errors in test.log, and this appears to mimic other tests that want to verify the run phase. Any ideas?

jlsherrill replied:
so i tested this out locally and i think the problem is that the IncrementalUpdate action generates output that is passed to the IncrementalUpdates action into its plan_self method:
https://github.com/Katello/katello/blob/master/app/lib/actions/katello/content_view/incremental_updates.rb#L20-L23
Here's where it passes to plan_self
https://github.com/Katello/katello/blob/master/app/lib/actions/katello/content_view/incremental_updates.rb#L36

From what i can gather, this isn't actually supported to do within a test, at least not in the way that was executed here.  You could try calling ForemanTasks.sync_task(), as we do that in other places in tests to fully run an action and I don't remember seeing that issue there?  

You could also stub The 'IncrementalUpdate' action and provide some fake output, we do that in some places like here:   https://github.com/Katello/katello/blob/master/test/actions/katello/repository_test.rb#L243-L245

jturel commented:
@jlsherrill since you were involved in some feedback+discussions on the original PR, can you take a look at the approach here? Jeremy will test it out end-to-end.

jlsherrill commented on app/lib/actions/katello/content_view/incremental_update/sync_and_install_content.rb:
```diff
@@ -0,0 +1,32 @@
+                    plan_action(::Actions::Katello::ContentView::CapsuleSync, content_view, environment)
+                  end
+                end
+              end
```
we may want to test what happens if a capsule can't be synced due to either:
a) it being down at the start of the sync
b) the sync failing for some reason after its started

I think for 
a) the whole task may abort which we probably don't want, we likely want best effort across all capsules
b) I *think* that the sync would be skipped and everything else would occur (Because CapsuleContent::Sync has a rescue_strategy of Skip), but we may want to double check that in reality.

jturel replied:
The incremental update completes with a warning but other syncs complete :+1:

jlsherrill commented:
One comment about verifying and handling failure scenarios, but otherwise this all looks good :+1:

jturel commented:
@jlsherrill  I changed the approach of this PR slightly. Rather than a new action, I modified `PromoteToEnvironment` (details above). Still need to test the behavior when capsules are offline but I'm running into something weird I'm hoping you can explain.

Even with the smart proxy being synced I can not install the applicable errata. I thought it had something to do with `skip_metadata_check` so I changed that to `true`. I thought forcing the metadata refresh would do the trick but no luck. However, when I perform a full sync of the FPC I can install the errata. Any ideas?

This is what the sync looks like through my incremental update

![image](https://user-images.githubusercontent.com/1979598/87713942-01991000-c779-11ea-823d-6f70dc68613e.png)

jturel commented:
[test katello]

chris1984 commented:
@jlsherrill are you reviewing this, or does it still need to be tested?

jlsherrill commented:
@chris1984 I'm happy to review the code if you want to test it!

chris1984 commented:
I tested this and hit an issue with the sync failing on the capsule, here is the info:

```bash
[root@client1 ~]# yum install walrus-0.71

=========================================================================================================================================================================================================================================================================================== Package                                                    Arch                                                       Version                                                       Repository                                                                                       Size
===========================================================================================================================================================================================================================================================================================Installing:
 walrus                                                     noarch                                                     0.71-1                                                        Default_Organization_Test-Product_Test-Repo                                                     2.4 k
Installing for dependencies:
 shark                                                      noarch                                                     0.1-1                                                         Default_Organization_Test-Product_Test-Repo                                                     2.4 k
 stork                                                      noarch                                                     0.12-2                                                        Default_Organization_Test-Product_Test-Repo                                                     2.4 k
 whale                                                      noarch                                                     0.2-1                                                         Default_Organization_Test-Product_Test-Repo                                                     2.4 k

Transaction Summary
===========================================================================================================================================================================================================================================================================================

[root@client1 ~]# rpm -qa | grep walrus
walrus-0.71-1.noarch

[root@client1 ~]# yum update walrus
Default_Organization_Test-Product_Test-Repo                                                                                                                                                                                                                         | 3.5 kB  00:00:00     
No packages marked for update
Uploading Enabled Repositories Report
```

Screenshot of host showing it needs the errata after doing the icremental update:

https://nimbusweb.me/s/share/4556155/y5t720mdik8xdiv7wn5q

Screenshot of my cv:

https://nimbusweb.me/s/share/4556162/gxpne9kkel9hgh7ko1z4

When running hammer I get the following:

[root@prtest ~]# hammer content-view version incremental-update --content-view-version-id=3 --errata-ids=RHEA-2012:0055 --update-all-hosts=1 --lifecycle-environment-ids=2
[..................................................................................................................................................................................................................................................................................] [100%]Content View: Test-View version 2.1
Added Content:
  Errata:
        RHEA-2012:0055
  Packages:
        penguin-0.9.1-1.noarch
        shark-0.1-1.noarch
        walrus-5.21-1.noarch
Error: A sub task failed

My capsule keeps failing on the sync with this error: "error"=>"Fail to update the existing erratum RHEA-2012:0056."},` traceback below:

```ruby
{"pulp_tasks"=>
  [{"exception"=>nil,
    "task_type"=>"pulp.server.managers.repo.sync.sync",
    "_href"=>"/pulp/api/v2/tasks/684df68b-fcb9-468d-9378-b0b848e111c9/",
    "task_id"=>"684df68b-fcb9-468d-9378-b0b848e111c9",
    "tags"=>
     ["pulp:repository:1-Test-View-Stage-68f3a9b0-2f7c-4537-afef-a4a6d0bc30a6",
      "pulp:action:sync"],
    "finish_time"=>"2020-08-19T19:43:48Z",
    "_ns"=>"task_status",
    "start_time"=>"2020-08-19T19:43:40Z",
    "traceback"=>
     "Traceback (most recent call last):\n" +
     "  File \"/usr/lib/python2.7/site-packages/celery/app/trace.py\", line 367, in trace_task\n" +
     "    R = retval = fun(*args, **kwargs)\n" +
     "  File \"/usr/lib/python2.7/site-packages/pulp/server/async/tasks.py\", line 686, in __call__\n" +
     "    return super(Task, self).__call__(*args, **kwargs)\n" +
     "  File \"/usr/lib/python2.7/site-packages/pulp/server/async/tasks.py\", line 108, in __call__\n" +
     "    return super(PulpTask, self).__call__(*args, **kwargs)\n" +
     "  File \"/usr/lib/python2.7/site-packages/celery/app/trace.py\", line 622, in __protected_call__\n" +
     "    return self.run(*args, **kwargs)\n" +
     "  File \"/usr/lib/python2.7/site-packages/pulp/server/controllers/repository.py\", line 855, in sync\n" +
     "    raise pulp_exceptions.PulpExecutionException(_('Importer indicated a failed response'))\n" +
     "PulpExecutionException: Importer indicated a failed response\n",
    "spawned_tasks"=>[],
    "progress_report"=>
     {"yum_importer"=>
       {"content"=>
         {"items_total"=>0,
          "state"=>"FINISHED",
          "error_details"=>[],
          "details"=>
           {"rpm_total"=>0, "rpm_done"=>0, "drpm_total"=>0, "drpm_done"=>0},
          "size_total"=>0,
          "size_left"=>0,
          "items_left"=>0},
        "comps"=>{"state"=>"NOT_STARTED"},
        "purge_duplicates"=>{"state"=>"NOT_STARTED"},
        "distribution"=>
         {"items_total"=>0,
          "state"=>"FINISHED",
          "error_details"=>[],
          "items_left"=>0},
        "modules"=>{"state"=>"FINISHED"},
        "errata"=>
         {"state"=>"FAILED",
          "error"=>"Fail to update the existing erratum RHEA-2012:0056."},
        "metadata"=>{"state"=>"FINISHED"}}},
    "queue"=>"reserved_resource_worker-1@proxy.katello.lan.dq2",
    "state"=>"error",
    "worker_name"=>"reserved_resource_worker-1@proxy.katello.lan",
    "result"=>nil,
    "error"=>
     {"code"=>"PLP0000",
      "data"=>{},
      "description"=>"Importer indicated a failed response",
      "sub_errors"=>[]},
    "_id"=>{"$oid"=>"5f3d80ebc3a0e3dc86237280"},
    "id"=>"5f3d80ebc3a0e3dc86237280"}],
 "poll_attempts"=>{"total"=>10, "failed"=>1}}
Exception:
Katello::Errors::PulpError: PLP0000: Importer indicated a failed response
Backtrace:
/opt/theforeman/tfm/root/usr/share/gems/gems/katello-3.18.0.pre.master/app/lib/actions/pulp/abstract_async_task.rb:121:in `block in external_task='
/opt/theforeman/tfm/root/usr/share/gems/gems/katello-3.18.0.pre.master/app/lib/actions/pulp/abstract_async_task.rb:119:in `each'
/opt/theforeman/tfm/root/usr/share/gems/gems/katello-3.18.0.pre.master/app/lib/actions/pulp/abstract_async_task.rb:119:in `external_task='
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.6/lib/dynflow/action/polling.rb:100:in `poll_external_task_with_rescue'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.6/lib/dynflow/action/polling.rb:22:in `run'
/opt/theforeman/tfm/root/usr/share/gems/gems/dynflow-1.4.6/lib/dynflow/action/cancellable.rb:14:in `run'
/opt/theforeman/tfm/root/usr/share/gems/gems/katello-3.18.0.pre.master/app/lib/actions/pulp/abstract_async_task.rb:45:in `run'
```

jlsherrill commented:
ACK from a code perspective.   @chris1984 that error looks unrelated to this pr, but it might be worth asking the pulp team about it

chris1984 commented:
ACK, works on 3.17.

After doing the steps, I can see the `walrus` package

```shell
[root@client ~]# rpm -qa | grep walrus
walrus-5.21-1.noarch
```

jturel commented:
rebased it correctly now

jturel commented:
[test katello]

jturel commented:
[test katello]

jturel commented:
[test katello]
